components/eamxx/src/share/field/field_utils_impl_dot.hpp

```diff
@@ -0,0 +1,387 @@
+#ifndef SCREAM_FIELD_UTILS_IMPL_DOT_HPP
+#define SCREAM_FIELD_UTILS_IMPL_DOT_HPP
+
+#include "ekat/kokkos/ekat_kokkos_utils.hpp"
+#include "ekat/mpi/ekat_comm.hpp"
+#include "share/field/field.hpp"
+
+namespace scream {
+
+// dot product of rank 1 field with a rank N field, return a rank N-1 field
+template <typename ST>
+Field do_dot_along_rank1_dim(const int &pd, const Field &f1, const Field &f2,
+                             const ekat::Comm *co) {
+  using KT          = ekat::KokkosTypes<DefaultDevice>;
+  using RangePolicy = Kokkos::RangePolicy<Field::device_t::execution_space>;
+  using TeamPolicy  = Kokkos::TeamPolicy<Field::device_t::execution_space>;
+  using TeamMember  = typename TeamPolicy::member_type;
+  using ESU         = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  const auto &l1 = f1.get_header().get_identifier().get_layout();
+
+  const auto &n2 = f2.get_header().get_identifier().name();
+  const auto &l2 = f2.get_header().get_identifier().get_layout();
+  const auto &u2 = f2.get_header().get_identifier().get_units();
+  const auto &g2 = f2.get_header().get_identifier().get_grid_name();
+
+  EKAT_REQUIRE_MSG(pd <= 5 && pd >= 0,
```

Comment:
I wonder if we should make this super generic. The impl below is quite long mostly b/c of handling different values of pd. What if we only supported pd=0 (for contraction along horiz dim) and pd=rank-1 (for contraction along vertical dim)? We could even make two separate methods, to make the code more self-explanatory:
```c++
Field horiz_contracttion (const Field& field, const Field& weight);
Field vert_contracttion (const Field& field, const Field& weight);
```
or something like that. We'd of course check that the input fields have the correct tag (COL for horiz and LEV/ILEV for vert), that it is at the expected, and that the extent match. We could also make it so that the `weight` field has JUST the horiz (resp. vert) dimension, so that it is indeed a weight. Maybe we should make weight optional too, so 
```c++
Field horiz_contraction (const Field& field, const Field* weight = nullptr);
```
I think this would cover most of our potential use cases.

Reply:
We backed the assumptions that "COL comes first and LEV/ILEV come last" in a lot of places, so I think it's safe to assume that will remain the case.

Reply:
> contract_field_along_lev (essentially as above, but pd = rank-1 = 2)

Well, LEV could be at pd=0 (for a (LEV) field), 1 (for a (COL,LEV) field) ,or 2 (for a (COL,CMP,LEV) field), so we cannot assume it's 2. But we _can_ assume it's rank-1.

Reply:
For reduction along CMP, assuming that the extent is "small", I would consider doing something like
```c++
Field out = f.get_component(0).clone(); // get output layout
out.deep_copy(0);
for (int icmp=0; icmp<num_cmp; ++icmp) {
  out.update(f.get_component(icmp),1,1);
}
```
Requires num_cmp loops, but they are ALL ||for, no ||reduce.

Reply:
I think so. I doubt we will ever need to reduce along a generic dimension. Our only two/three cases are COL, LEV/ILEV, and CMP. The latter is prob easier to impl via for loop and calls to `update`; the COL case can assume it's the 1st dim, and the LEV/ILEV case can assume it's the last dim. And we can assume we contract a single field, _possibly_ weighted by a 1d field (of length equal to the extent along which we are contracting).

Reply:
As you said, these use cases will cover our most likely usage: horizontal sum/avg, vertical sum/avg, and reduction across bands (or other CMP extents).